File: src/main/java/renankummer/udemy/springwebapp/model/Book.java
package renankummer.udemy.springwebapp.model;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import javax.persistence.*;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;
import java.util.UUID;

@Entity
@Data @NoArgsConstructor @AllArgsConstructor @Builder
public class Book {

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private Long id;

    @Column(name = "title", nullable = false)
    private String title;
    @Column(name = "isbn", nullable = false)
    @Builder.Default
    private String isbn = UUID.randomUUID().toString();

    @ManyToMany
    @JoinTable(
            name = "author_book",
            joinColumns = @JoinColumn(name = "book_id"),
            inverseJoinColumns = @JoinColumn(name = "author_id")
    )
    @Builder.Default
    private Set<Author> authors = new HashSet<>();

    @ManyToOne
    private Publisher publisher;

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Book book = (Book) o;
        return Objects.equals(id, book.id);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }
}

File: src/main/kotlin/renankummer/udemy/springwebapp/model/Book.kt
package renankummer.udemy.springwebapp.model

import java.util.*
import javax.persistence.*
import kotlin.collections.HashSet

@Entity
@Table(name = "books")
data class Book (
        @Id
        @GeneratedValue(strategy = GenerationType.AUTO)
        val id: Long = 0,

        @Column(name = "title", nullable = false)
        var title: String,
        @Column(name = "isbn", nullable = false)
        var isbn: String = UUID.randomUUID().toString(),

        @ManyToMany
        @JoinTable(
                name = "authors_books",
                joinColumns = [ JoinColumn(name = "books_id", nullable = false) ],
                inverseJoinColumns = [ JoinColumn(name = "authors_id") ]
        )
        var authors: Set<Author> = HashSet<Author>()
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Book

        if (id != other.id) return false

        return true
    }

    override fun hashCode() = id.hashCode()
}
